### DISCUSSION TITLE
Router unexpectedly slow

### DISCUSSION THREAD

**Discussion by andraaspar** [solidjs/solid#2208] [General]
In the following sample code (which does not work in the Playground for some reason), waiting for Page A to load **and then** switching to Page B is unexpectedly slow. All other pages work fine. It seems to suspend before the navigation could happen.

The workaround is commented out in AppComp.

Is this a bug, or am I doing something wrong? Is there a better solution?

```ts
import { A, HashRouter, Route, RouteSectionProps } from '@solidjs/router'
import { Suspense, createResource, Show, type JSX } from 'solid-js'
import { render } from 'solid-js/web'

function AppComp() {
  return (
    <>
      <style>{`
      .nav.active {
      outline: 3px solid crimson;
    }`}</style>
      <HashRouter root={PageWrapperComp}>
        <Route path={'/'} component={PageXComp} />
        <Route
          path={'/page/:name'}
          component={PageXComp}
          // component={(props) => (
          // 	<Show keyed when={JSON.stringify(props)}>
          // 		<PageXComp {...props} />
          // 	</Show>
          // )}
        />
        <Route path='/another' component={AnotherPageComp} />
      </HashRouter>
    </>
  )
}

function PageWrapperComp(props: { children?: JSX.Element }) {
  return (
    <>
      <p>
        Please wait for the page to load completely before clicking any link.
      </p>
      <div>
        Navigation:{' '}
        <A href='/' end class='nav'>
          Root
        </A>{' '}
        <A href='/page/a' class='nav'>
          Page A
        </A>{' '}
        <A href='/page/b' class='nav'>
          Page B
        </A>{' '}
        <A href='/another' class='nav'>
          Another
        </A>
      </div>
      <div>Content:</div>
      {props.children}
    </>
  )
}

function PageXComp(props: RouteSectionProps) {
  console.log(JSON.stringify(props))
  function getPageName() {
    return props.params.name?.toLocaleUpperCase() ?? 'Root'
  }
  const [resource] = createResource(getPageName, (pageName) => {
    return new Promise<string>((resolve) => {
      setTimeout(() => {
        resolve(`Hi ${pageName}!`)
      }, 3000)
    })
  })
  return (
    <Suspense fallback={'Loading PageXComp...'}>
      <div>This is PageXComp: {resource()}</div>
    </Suspense>
  )
}

function AnotherPageComp() {
  const [resource] = createResource(() => {
    return new Promise<string>((resolve) => {
      setTimeout(() => {
        resolve('Hello!')
      }, 3000)
    })
  })
  return (
    <Suspense fallback={'Loading AnotherPageComp...'}>
      <div>This is AnotherPageComp: {resource()}</div>
    </Suspense>
  )
}

render(() => <AppComp />, document.getElementById('app')!)

```

**Comment by andraaspar** [comment]
You can see it live [here](https://codesandbox.io/p/devbox/router-unexpectedly-slow-2208-fh4dqp?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cly5negi500072a699zvnkaqj%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cly5negi500022a695mmn1xjb%2522%253A%257B%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cly5negi500062a69j3171acd%2522%253A%257B%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500052a69ka127cld%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2523%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500052a69ka127cld%2522%257D%252C%2522cly5negi500042a69n32xanhb%2522%253A%257B%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500032a6934c12jh9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500032a6934c12jh9%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

**Comment by kurrestahlberg** [comment]
I've seen the exact same thing. Rendering the new data is completely delayed until the async function in `createResource()` completes. If I swap the source from `createResource()` for a `createEffect()` that calls `refresh()` it works exactly how I imagine it should: Other elements are rendered immediately, suspense gets triggered, etc

Changing your example like below should fix it, even though I don't think this is what it should be

```typescript
const [resource, {refetch}] = createResource(() => {
    const pageName = getPageName()
    return new Promise<string>((resolve) => {
      setTimeout(() => {
        resolve(`Hi ${pageName}!`)
      }, 3000)
    })
  })

createEffect(() => {
     refetch(getPageName())
})
```

Does anyone know what would be the correct way to fix this? Is it possibly a bug in `createResource()` or is there just something I've misunderstood?

**Comment by andraaspar** [comment]
I’d been waiting for a response here for ages, but it never came. My solution was to ditch solidjs/router and never use suspend. Those two features are shaky. Without them, solid is reliable. So long as you don’t accidentally create an infinite effect loop. That is another area that could use some improvement.

**Comment by ryansolid** [comment]
This is by design. The router bakes in Transitions which mirror paint holding of routing which rendering offscreen. There is an isRouting hook to provide affordances. You can also nest Suspense boundaries as new ones will not participate in Transitions.

You can see more on Transistions here:
https://www.solidjs.com/tutorial/async_transitions

**Comment by andraaspar** [comment]
Exactly. So the router (and suspense) is suitable for SSR, but broken in client-side-only scenarios. React is having similar issues around suspense, I hear. I’ve been holding off for the WIP suspense primitives I read about in some issue around here. Never heard about nesting suspense boundaries though. How would that solve the issue [here](https://codesandbox.io/p/devbox/router-unexpectedly-slow-2208-fh4dqp?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cly5negi500072a699zvnkaqj%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cly5negi500022a695mmn1xjb%2522%253A%257B%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cly5negi500062a69j3171acd%2522%253A%257B%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500052a69ka127cld%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2523%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500052a69ka127cld%2522%257D%252C%2522cly5negi500042a69n32xanhb%2522%253A%257B%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500032a6934c12jh9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500032a6934c12jh9%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)?

**Edit:** I tried to nest suspense boundaries [like this](https://codesandbox.io/p/devbox/router-unexpectedly-slow-2208-forked-m9r6k8?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cly5negi500072a699zvnkaqj%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cly5negi500022a695mmn1xjb%2522%253A%257B%2522id%2522%253A%2522cly5negi500022a695mmn1xjb%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522cly5negi500062a69j3171acd%2522%253A%257B%2522id%2522%253A%2522cly5negi500062a69j3171acd%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500052a69ka127cld%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2523%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500052a69ka127cld%2522%257D%252C%2522cly5negi500042a69n32xanhb%2522%253A%257B%2522id%2522%253A%2522cly5negi500042a69n32xanhb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cly5negi500032a6934c12jh9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cly5negi500032a6934c12jh9%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D), but it did not work.

```tsx
        <Route
          path={"/page/:name"}
          component={PageXWrapperComp}
        />

// ...

function PageXWrapperComp(props: RouteSectionProps) {
  return (
    <Suspense fallback={"Loading PageXComp inside PageXWrapperComp..."}>
      <PageXComp {...props} />
    </Suspense>
  );
}
```

Going from Page A to Page B is still slow, and does not show the suspense fallback message. How did you mean that?

**Comment by andraaspar** [comment]
Thissss dissssscusssssionnnn issss rattttttherrrrr sloowwwwww.
